rfc/graph-pruning-garbage-collection.md

```diff
@@ -0,0 +1,147 @@
+
+It's also possible to disable the rules:
+
+```shell
+npx wgc namespace update production --unused-fields-warn=0 --unused-fields-fail=0
+```
+
+Maybe you'd only like to have warnings:
+
+```shell
+npx wgc namespace update production --unused-fields-warn=30 --unused-fields-fail=0
+```
+
+### The @deprecated directive resets the counter
+
+If an unused field causes a warning or a failure in the CI/CD pipeline,
+you can deprecate the field to reset the counter:
+
+```graphql
+type Query {
+  oldField: String @deprecated(reason: "This field is deprecated")
+}
+```
+
+Let's say that the `oldField` is unused for 30 days.
+The CI/CD pipeline will show a warning.
+We deprecate the field, and the counter is reset.
+If the deprecated field is unused for another 30 days,
+the CI/CD pipeline will show a warning again.
+
+The field now was unused for 60 days,
+which should be enough time to be sure that the field is not used anymore.
+
+### Analytics must exist for more than the threshold period
+
+To avoid false positives, we should only consider fields that have been unused for more than the threshold period.
+This means that we need to keep track of the last time a field was changed.
+When a field is added or modified, we keep track of the timestamp.
+Fields that have been added or modified within the threshold period are not considered for pruning.
+
+Let's say that the threshold period is 30 days.
+If a field was added 20 days ago and is not used,
+it should not be considered for pruning.
+A field must exist for more than 30 days in the Schema Registry to be considered for pruning.
+
+### Notifications
+
+Internally, Cosmo can run a daily job to check for unused fields.
+If a warning or failure is detected,
+a notification can be sent to the Slack channel or via Webhook.
+
+You can configure the notification settings at the organization level.
+
+### Defaults
+
+All existing namespaces and newly created namespaces will have the following default pruning rules:
+
+1. Warn after 15 days
+2. Fail after 30 days
+
+These should be good defaults for most organizations to keep the schema clean.
+
+## Why structuring the workflow like this?
+
+By using the existing CI/CD pipeline, we're giving the Subgraph owners direct feedback on the quality of their schema.
+Instead of having a separate tool or process to clean up the schema,
+we're integrating this into the existing workflow.
+
+We've also considered of suggesting Pull Requests to remove unused fields automatically,
+but we think that we'd like to give the Subgraph owners the control over their schema.
+We also don't want to introduce additional workflows that might be too intrusive.
+
+## Escape Hatch
+
+It's possible that some fields are not used in the Schema but there are other reasons to keep them,
+e.g. for business reasons or because there are plans to use them in the future.
+
+In this case, you can add a special comment to the field:
+
+```graphql
+type Query {
+  # wgc-check:ignore-unused
```

Comment:
I believe this could be challenging in workflows where the schema is an artifact generated from the code.

I'd prefer if we could implement ignore rules based on tag associated with fields or entire types

Reply:
How are you generating your Schema? Are you using a code-first approach? Do you think a directive could work better? E.g. like this:

```graphql
type Query {
  oldField: String @keep(reason: "This field will be used in the future")
}
```

Or we put the directive on the type and it affects all enclosing fields:

```graphql
type Query @keep(reason: "This field will be used in the future") {
  oldField: String
}
```

Reply:
Currently we are using code-first approach with [nestjs](https://docs.nestjs.com/graphql/quick-start#code-first) and [graphql-kotlin](https://opensource.expediagroup.com/graphql-kotlin/docs/).

Great idea! A dedicated directive could be a nice option too.

2c: "keep" is a bit too ambiguous and also too specific to my taste 😅 . I wonder if there is a possibility to create more generic directive to override linter checks in general. Like `@SuppressWarning` in java.

Reply:
Based on other feedback, I'm leaning towards adding this feature into the linter category. If we're doing that, we could do a more generic approach like this:

```graphql
type Query {
  oldField: String @ignoreLint(reason: "This field will be used in the future")
}
```

What are your thoughts on this?